cmd/panasas/config/client.go

```diff
@@ -0,0 +1,389 @@
+}
+
+func (e ErrUnexpectedContentType) Error() string {
+	return fmt.Sprintf("Unexpected HTTP content type: %q", string(e))
+}
+
+// PanasasConfigClient represents a Panasas config agent client
+type PanasasConfigClient struct {
+	agentHost  string
+	agentPort  uint
+	httpClient *http.Client
+
+	configRevision *string
+}
+
+// NewPanasasConfigClient returns a configured PanasasConfigClient
+func NewPanasasConfigClient(host string, port uint) *PanasasConfigClient {
+	client := PanasasConfigClient{
+		agentHost:  host,
+		agentPort:  port,
+		httpClient: &http.Client{},
+	}
+	return &client
+}
+
+func (c *PanasasConfigClient) getPanasasConfigStoreURL(base string, elem ...string) (*url.URL, error) {
+	var path string = base
+	var e error
+
+	for _, eachElem := range elem {
+		path, e = url.JoinPath(base, url.PathEscape(eachElem))
+		if e != nil {
+			return nil, e
+		}
+	}
+
+	return url.Parse(
+		fmt.Sprintf("http://%v:%v%v", c.agentHost, c.agentPort, path),
+	)
+}
+
+func (c *PanasasConfigClient) extractConfigRevision(resp *http.Response) {
+	revisionHeaders := resp.Header.Values(panasasHTTPS3ConfigRevisionHeader)
+	if len(revisionHeaders) > 0 {
+		var revision string = revisionHeaders[0]
+		c.configRevision = &revision
+	}
+}
+
+// closeResponseBody close non nil response with any response Body.
+// convenient wrapper to drain any remaining data on response body.
+//
+// Subsequently this allows golang http RoundTripper
+// to re-use the same connection for future requests.
+// (copied from minio-go)
+func closeResponseBody(resp *http.Response) {
+	// Callers should close resp.Body when done reading from it.
+	// If resp.Body is not closed, the Client's underlying RoundTripper
+	// (typically Transport) may not be able to re-use a persistent TCP
+	// connection to the server for a subsequent "keep-alive" request.
+	if resp != nil && resp.Body != nil {
+		// Drain any remaining Body and then close the connection.
+		// Without this closing connection would disallow re-using
+		// the same connection for future uses.
+		//  - http://stackoverflow.com/a/17961593/4465767
+		io.Copy(ioutil.Discard, resp.Body)
+		resp.Body.Close()
+	}
+}
+
+// PanasasObjectList represents a list of objects as returned by configuration agent
+type PanasasObjectList struct {
+	Objects []string
+}
+
+// GetObjectsList returns a list of objects
+func (c *PanasasConfigClient) GetObjectsList(prefix string) ([]string, error) {
+	u, err := c.getPanasasConfigStoreURL("/objects")
+
+	if err != nil {
+		fmt.Printf("Failed formatting URL for /objects with prefix %q\n", prefix)
+		log.Fatal(err)
+		return []string{}, err
+	}
+
+	req := &http.Request{
+		Method:        http.MethodGet,
+		URL:           u,
+		Proto:         "HTTP/1.1",
+		ProtoMajor:    1,
+		ProtoMinor:    1,
+		Header:        make(http.Header),
+		Host:          u.Host,
+		ContentLength: 0,
+	}
+
+	req.Header.Set(panasasHTTPS3PrefixHeader, prefix)
+
+	resp, err := c.httpClient.Do(req)
```

Comment:
I think it makes sense to move the common part into a separate method, and then use it inside the `GetObjectsList`,  `GetObject` ... methods, since there is a lot of the same code here.

I mean this one

```
u, err := c.getPanasasConfigStoreURL("/objects")

	if err != nil {
		fmt.Printf("Failed formatting URL for /objects with prefix %q\n", prefix)
		log.Fatal(err)
		return []string{}, err
	}

	req := &http.Request{
		Method:        http.MethodGet,
		URL:           u,
		Proto:         "HTTP/1.1",
		ProtoMajor:    1,
		ProtoMinor:    1,
		Header:        make(http.Header),
		Host:          u.Host,
		ContentLength: 0,
	}

	req.Header.Set(panasasHTTPS3PrefixHeader, prefix)

	resp, err := c.httpClient.Do(req)
```